pkg/client/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,219 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package leaderelection implements leader election of a set of endpoints.
+// It uses an annotation in the endpoints object to store the record of the
+// election state.
+package leaderelection
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/client/record"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/wait"
+)
+
+const (
+	JitterFactor = 1.2
+
+	LeaderElectionRecordAnnotationKey = "control-plane.alpha.kubernetes.io/leader"
+)
+
+func NewLeaderElector(lec LeaderElectionConfig) (*LeaderElector, error) {
+	if lec.LeaseDuration < lec.RenewDeadline {
+		return nil, fmt.Errorf("leaseDuration must be greater than renewDeadline")
+	}
+	if lec.RenewDeadline < time.Duration(JitterFactor*float64(lec.RetryPeriod)) {
+		return nil, fmt.Errorf("renewDeadline must be greater than retryPeriod*JitterFactor")
+	}
+	return &LeaderElector{
+		config: lec,
+	}, nil
+}
+
+type LeaderElectionConfig struct {
+	EndpointsMeta api.ObjectMeta
+	Identity      string
+
+	Client        client.Interface
+	EventRecorder record.EventRecorder
+
+	LeaseDuration time.Duration
+	RetryPeriod   time.Duration
+	RenewDeadline time.Duration
+
+	Callbacks LeaderCallbacks
+}
+
+type LeaderCallbacks struct {
+	// Async callbacks triggered on leader change
+	OnStartedLeading func(stop <-chan struct{})
+	OnStoppedLeading func()
+}
+
+// possible future async callbacks:
+//  * OnChallenge()
+//  * OnNewLeader()
+// possible future synchronous calls:
+//  * IsLeader()
+//  * GetLeader()
+type LeaderElector struct {
+	config LeaderElectionConfig
+	// internal bookkeeping
+	observedRecord LeaderElectionRecord
+	observedTime   time.Time
+}
+
+// TODO(mikedanese): this should potentially be versioned
+type LeaderElectionRecord struct {
+	HolderIdentity       string           `json:"holderIdentity"`
+	LeaseDurationSeconds int              `json:"leaseDurationSeconds"`
+	AquireTime           unversioned.Time `json:"aquireTime"`
+	RenewTime            unversioned.Time `json:"renewTime"`
+}
+
+// Run starts the leader election loop
+func (le *LeaderElector) Run() {
+	defer func() {
+		util.HandleCrash()
+		le.config.Callbacks.OnStoppedLeading()
+	}()
```

Comment:
seems like somewhat duplicated effort given the `stop` chan, which could signal the same thing (if it was closed via `defer` as I suggested). the difference being that a client won't have a reference to `stop` unless they've received the start callback.

I think the API would be more clear if "started" and "stopped" were NOT callbacks, but instead were exposed as latches:
```go
// Started returns a chan that closes upon being elected for leadership. No objects are ever sent over the chan.
func (le *LeaderElector) Started() <-chan struct{} {... }
// Stopped returns a chan that closes upon being removed from leadership. No objects are ever sent over the chan.
func (le *LeaderElector) Stopped() <-chan struct{} {... }
```
.. and get rid of callbacks all together.